src/Pact/Eval.hs

```diff
@@ -205,52 +225,136 @@ loadModule m bod1 mi g0 = do
-                  (Nothing,_) -> evalError (_nInfo f) ("Cannot resolve \"" ++ show f ++ "\"")
-        return (d',dn,mapMaybe (either Just (const Nothing)) $ toList d')
-  sorted <- forM cs $ \c -> case c of
-              AcyclicSCC v -> return v
-              CyclicSCC vs ->
-                evalError (if null vs then mi else _tInfo $ view _1 $ head vs) $ "Recursion detected: " ++ show vs
-  let defs :: HM.HashMap Text Ref
-      defs = foldl dresolve HM.empty sorted
-      -- insert a fresh Ref into the map, fmapping the Either to a Ref via 'unify'
-      dresolve ds (d,dn,_) = HM.insert dn (Ref (fmap (unify ds) d)) ds
-      evalConstRef r@Ref {} = runPure $ evalConsts r
-      evalConstRef r@Direct {} = runPure $ evalConsts r
-  evaluatedDefs <- traverse evalConstRef defs
-  installModule (m,evaluatedDefs)
-  (evalRefs.rsNew) %= ((_mName m,(m,evaluatedDefs)):)
-  return (g1,modDefs1)
-
-
-
+  evaluatedDefs <- evaluateDefs mi modDefs1
+  solvedDefs <- evaluateConstraints mi m evaluatedDefs
+  let md = ModuleData m solvedDefs
+  installModule md 
+  (evalRefs . rsNewModules) %= HM.insert _mName md
+  return (g1, modDefs1)
+loadModule i@Interface{..} body info gas0 = do
+  (gas1,idefs) <- case instantiate' body of
+    (TList bd _ _bi) -> do
+      let doDef (g,rs) t = do
+            dnm <- case t of
+              TDef {..} -> return $ Just _tDefName
+              TConst {..} -> return $ Just $ _aName _tConstArg
+              TSchema {..} -> return $ Just $ asString _tSchemaName
+              TUse {..} -> evalUse _tModuleName _tModuleHash _tInfo >> return Nothing
+              _ -> evalError (_tInfo t) "Invalid interface member"
+            case dnm of
+              Nothing -> return (g, rs)
+              Just dn -> do
+                g' <- computeGas (Left (_tInfo t,dn)) (GModuleMember i)
+                return (g + g',(dn,t):rs)
+      second HM.fromList <$> foldM doDef (gas0,[]) bd
+    t -> evalError (_tInfo t) "Malformed interface"
+  evaluatedDefs <- evaluateDefs info idefs
+  let md = ModuleData i evaluatedDefs
+  installModule md
+  (evalRefs . rsNewModules) %= HM.insert _interfaceName md
+  return (gas1, idefs)
+
+evaluateDefs :: Info -> HM.HashMap Text (Term Name) -> Eval e (HM.HashMap Text Ref)
+evaluateDefs info defs = do
+  cs <- traverseGraph defs
+  sortedDefs <- detectCycles cs
+  unifiedDefs <- foldl dresolve (pure HM.empty) sortedDefs
+  traverse evalConstRef unifiedDefs
+  where
+    dresolve eds (d,dn,_) = do
+      ds <- eds
+      ud <- traverse (unify info ds) d
+      return $ HM.insert dn (Ref ud) ds
+    evalConstRef = runPure . evalConsts
+    detectCycles sccs = forM sccs $ \c ->
+      case c of
+        AcyclicSCC v -> return v
+        CyclicSCC vs -> evalError (if null vs then info else _tInfo $ view _1 $ head vs) $
+          "Recursion detected: " ++ show vs
+
+traverseGraph :: HM.HashMap Text (Term Name) -> Eval e [SCC (Term (Either Text Ref), Text, [Text])]
```

Comment:
Maybe we should introduce a type synonym for the `Either Text Ref`, use it here and in `unify`, and comment the type synonym to document the intention of this (which IIRC is Text is a to-be-resolved, Ref is resolved)

Reply:
I'll do that

Reply:
nm, docs are already sufficient